### PR TITLE
ROX-10966 local-sensor: record and replay k8s events

### PR DESCRIPTION
**Pull request by lvalerom**
## Description

This PR aims to enable a way to record k8s events handled by sensor as well as a way to replay those events.

## Checklist
- [x] Investigated and inspected CI test results
- ~[ ] Unit test and regression tests added~
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

Local testing performed with `local-sensor`

```bash
go run tools/local-sensor/main.go -record -record-out=trace.jsonl
go run tools/local-sensor/main.go -replay -replay-in=trace.jsonl -delay=0s -resync=5s
```

### CODE REVIEW

File: sensor/debugger/k8s/trace.go
Comment (fredrb) [inline sensor/debugger/k8s/trace.go:68]: Why are we creating a folder on _reading_ the trace file? If you just want to check if the file exists you can do `os.Stat`

File: sensor/debugger/k8s/trace.go
Comment (fredrb) [inline sensor/debugger/k8s/trace.go:89]: Why do we want to read the file line-by-line? Isn't it easier to read the whole thing? `os.ReadFile` is a utility function that handles the reading buffer for you. Even if the file is MBs large, we won't read more than one, right? 
It would simplify a lot the logic here.

File: sensor/kubernetes/listener/resources/dispatcher.go
Comment (fredrb) [reply sensor/kubernetes/listener/resources/dispatcher.go:199]: If this is going to be a thing, I mean. Adding more of these wrappers would be annoying.

File: tools/local-sensor/main.go
Comment (fredrb) [inline tools/local-sensor/main.go:75]: nit: `s/fsc/sensorConfig`

File: tools/local-sensor/main.go
Comment (fredrb) [inline tools/local-sensor/main.go:168]: Could you please handle the error here?

File: tools/local-sensor/main.go
Comment (fredrb) [inline tools/local-sensor/main.go:166]: Why are you guarding this by an `Enabled` flag and not an if statement here? Wouldn't it be simpler to:
```golang
var traceRec k8s.TraceWriter
if fsc.ShallRecordK8sInput {
        traceRec = &k8s.TraceWriter{
		Destination: path.Clean(fsc.RecordK8sFile),
        }
        _ = traceRec.Init()
}
```

File: tools/local-sensor/main.go
Comment (fredrb) [inline tools/local-sensor/main.go:66]: nit: not a huge fan of prefixing these with `Shall*`. Maybe simply `ReplayK8sEnabled`?

**Review by fredrb — CHANGES_REQUESTED** [review]
First pass done. Didn't test it myself yet. Let me know what you think about the suggestions, mainly the bigger one about using reflection and I'll continue reviewing later. I think it has some potential to simplify things here.

Conceptually, I think this is great, my comments are more from the code design perspective. This is a great step toward our brand new sensor test framework. 🙂

File: sensor/debugger/k8s/fake.go
Comment (lvalerom) [reply sensor/debugger/k8s/fake.go]: Great suggestion! I've changed the code to use reflection. Let me know what you think.

File: sensor/debugger/k8s/trace.go
Comment (lvalerom) [reply sensor/debugger/k8s/trace.go]: Done!

File: sensor/debugger/k8s/trace.go
Comment (lvalerom) [reply sensor/debugger/k8s/trace.go]: oops.. I think this was a mistake from copy-pasting from the writer 😅

File: sensor/debugger/k8s/trace.go
Comment (lvalerom) [reply sensor/debugger/k8s/trace.go]: The initial idea was to have a generic reader that implemented `io.Reader` and pass it to `sensor` but it morphed into this. I simplified the code with your suggestion. Let me know what you think.

File: tools/local-sensor/main.go
Comment (lvalerom) [reply tools/local-sensor/main.go]: Done

File: tools/local-sensor/main.go
Comment (lvalerom) [reply tools/local-sensor/main.go]: oops! Done

File: sensor/kubernetes/listener/resources/dispatcher.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/dispatcher.go]: Yeah. I think it's much more convenient your way.

File: tools/local-sensor/main.go
Comment (lvalerom) [reply tools/local-sensor/main.go]: I got rid of the Enable field in both `TraceWriter` and `TraceReader` I think we don't need them (at least for the time being)

File: sensor/debugger/k8s/fake.go
Comment (vikin91) [inline sensor/debugger/k8s/fake.go:213]: Following on my previous comment, this function is a writer so it should create the channel inside and return it to the caller.

File: sensor/debugger/k8s/fake.go
Comment (vikin91) [inline sensor/debugger/k8s/fake.go:231]: Going further here, this function should also create a channel, multiplex the channels returned by `handleRunOp` and return the channel containing all messages to the caller

File: sensor/debugger/k8s/trace.go
Comment (vikin91) [inline sensor/debugger/k8s/trace.go:58]: This is fine for now but we need to make sure to produce a meaningful error in case the file is too big.

File: sensor/debugger/k8s/fake.go
Comment (vikin91) [reply sensor/debugger/k8s/fake.go:160]: (read this comment after reading my other comments) 

Here we have the following possibilities:
1. Close the channel as suggested and see whether there are any issues. If any panics appear we need to then implement the concurrency patterns as I suggested in other comments.
2. Not close the channel and leave it as it is. In this particular case there should be no severe consequences of this. I think it might be fine leaving it like this as this coat will not be run in the production.

**Review by vikin91 — COMMENTED** [review]
This looks really nice thanks for finishing it. I still haven't run the code but I will do this on Monday. I will review green or red after I play with the binary.

Please have a look at the issue with concurrency that I mentioned and tell me what you think about it.

**Comment by vikin91** [comment]
I think that the https://github.com/stackrox/stackrox/pull/1799 is now obsolete and can be closed, right?

File: sensor/debugger/k8s/fake.go
Comment (lvalerom) [reply sensor/debugger/k8s/fake.go:168]: I think we can just close the channel in the `CreateEvents` function. 

```golang
ch := make(chan string)
f.done = concurrency.NewSignal()
objs, err := f.Reader.ReadFile()
if err != nil {
    return err
}
f.Init()
go func() {
    defer close(ch)
```

**Comment by lvalerom** [comment]
> I think that the #1799 is now obsolete and can be closed, right?

Yes, you can close #1799.

File: sensor/debugger/k8s/fake.go
Comment (fredrb) [inline sensor/debugger/k8s/fake.go:164]: Wouldn't `case <-done.WaitC()` work where?

File: sensor/debugger/k8s/fake.go
Comment (fredrb) [inline sensor/debugger/k8s/fake.go:188]: The `length` of a channel isn't the same thing as the length of a slice. What you're defining here is a `buffered` channel that allows for sending data into a channel in a non-blocking way up until the buffer size. See [this as an example](https://go.dev/tour/concurrency/3).

I think it's fine to have a buffered channel and even more optimal here, but you have to consider this when handling the `done` signal. See my next comment.

File: sensor/debugger/k8s/fake.go
Comment (fredrb) [inline sensor/debugger/k8s/fake.go:166]: I think this might lead to a race condition. 

If you're using a buffered channel, it means that the emitter of events (goroutine under `CreateEvents`) is allowed to write in a non-blocking way. Therefore, it could be that once the done signal is called, `ch` has more than one record waiting in the queue. So if we're only reading one here is not sufficient. You'd still have to loop until the channel is empty (or closed).

Comment (fredrb) [reply sensor/debugger/k8s/fake.go:166]: I think it might make more sense to control the flow by simply iterating on the channel until closed rather than using two constructs (a channel and a signal).

**Review by fredrb — CHANGES_REQUESTED** [review]
Thanks for making this reflection based. I think it's much cleaner now. I think there still might be some concurrency problems after you implemented @vikin91's suggestion.

File: sensor/debugger/k8s/fake.go
Comment (lvalerom) [reply sensor/debugger/k8s/fake.go]: Yeah, it was a work in progress. After my chat with @vikin91 we changed this part a lot.

File: sensor/debugger/k8s/fake.go
Comment (lvalerom) [reply sensor/debugger/k8s/fake.go]: You're right, it could lead to a race condition. We changed the code in this part too after the chat.

File: sensor/debugger/k8s/fake.go
Comment (lvalerom) [reply sensor/debugger/k8s/fake.go]: I believe so. I've changed that in the last commit.

File: sensor/debugger/k8s/fake.go
Comment (lvalerom) [reply sensor/debugger/k8s/fake.go:216]: Just to keep record here: After our chat we reworked all the concurrency and it is significantly different now.

File: sensor/debugger/k8s/trace.go
Comment (lvalerom) [reply sensor/debugger/k8s/trace.go:58]: I think we can leave it like this for now. If at some point we need really large files we can always rework this part.

File: sensor/debugger/k8s/fake.go
Comment (fredrb) [reply sensor/debugger/k8s/fake.go]: Ok. Let me know when it's ready to reviewed again.

**Comment by lvalerom** [comment]
> I am getting
> 
> ```
> # github.com/stackrox/rox/sensor/debugger/k8s
> sensor/debugger/k8s/fake.go:14:2: invalid import "github.com/stackrox/rox/sensor/kubernetes/listener/resources": sensor/debugger cannot import from sensor/kubernetes/listener/resources; only allowed roots are [sensor/debugger generated image pkg]
> ```
> 
> when running `make style`. Could you check please?

That's weird, that should be fixed with the changes I added in the `analyzer`. Did you fetched the last changes? I've just tried `make style` in a fresh cloned repository and it seems to be working.

**Review by vikin91 — APPROVED** [review]
I am unblocking this, as we worked on this together, but please double-check it before merging.

I have run this, but was unable to see sensor stop on its own after all events were replayed - is this intended?

**Comment by vikin91** [comment]
> > I am getting
> > ```
> > # github.com/stackrox/rox/sensor/debugger/k8s
> > sensor/debugger/k8s/fake.go:14:2: invalid import "github.com/stackrox/rox/sensor/kubernetes/listener/resources": sensor/debugger cannot import from sensor/kubernetes/listener/resources; only allowed roots are [sensor/debugger generated image pkg]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > when running `make style`. Could you check please?
> 
> That's weird, that should be fixed with the changes I added in the `analyzer`. Did you fetch the last changes? I've just tried `make style` in a fresh cloned repository and it seems to be working.

If it works for you, then all is good :)

File: sensor/debugger/k8s/fake.go
Comment (lvalerom) [reply sensor/debugger/k8s/fake.go]: This was supposed to be used in the tests.

The idea behind `WaitForDone` and `WaitForMinimumResources` was to encapsulate the handling of reading the channel in the `FakeEventsManager` but after a little bit of thinking I think I'm going with the approach we took this morning and just handle the channels outside.

File: sensor/debugger/k8s/fake.go
Comment (lvalerom) [reply sensor/debugger/k8s/fake.go]: Yeah. I think it's simpler to just return them. I've changed the code accordingly.

File: tools/local-sensor/main.go
Comment (fredrb) [inline tools/local-sensor/main.go:232]: If you decide to implement the suggestion above, you could handle the error signal here other than in a separate `goroutine` like it's being done now.

```golang
select {
   case <-time.Tick(localConfig.Duration):
      // ran scenario for long enough, continue
   case <- errSig.Done():
      // something went wrong 
}
```

**Review by fredrb — APPROVED** [review]
Thanks for working on the changes, Luis. I'm unblocking this as well, but I still think there are some potential improvements in the error handling. Consider using the constructs in our `concurrency` package to make the code more maintainable. 

If you agree with the suggestion but don't want to do this now, then could you please add a tech debt JIRA task in our local-sensor epic for this?

File: tools/local-sensor/main.go
Comment (lvalerom) [reply tools/local-sensor/main.go:232]: Makes sense. I will do this in the new task.

**Comment by lvalerom** [comment]
> Thanks for working on the changes, Luis. I'm unblocking this as well, but I still think there are some potential improvements in the error handling. Consider using the constructs in our `concurrency` package to make the code more maintainable.
> 
> If you agree with the suggestion but don't want to do this now, then could you please add a tech debt JIRA task in our local-sensor epic for this?

I agree with your suggestion. I think we can do that in a different PR so we can unblock this. There are a couple enhancements that need to be done either way for testing the re-sync. I will open follow up PRs for those as well.

**Comment by lvalerom** [comment]
> @lvalerom create a ticket in local sensor epic also, please :-)

Here [ROX-11396](https://issues.redhat.com/browse/ROX-11396) :)

**Pull request merged**
